### ISSUE TITLE
[Feature Request]: Support "Warning" for Validation Webhook

### ISSUE THREAD

**Issue by STRRL** [kubernetes-sigs/controller-runtime#1896]
refer: https://github.com/kubernetes-sigs/controller-runtime/issues/1788

After Kubernetes 1.19, the webhook could respond with "warning"s, refer:  https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#response

Warnings for validation webhook would bring much help for users with interacting with API.

Currently, users of controller-runtime could implement their "customized" `admission.Handler` to respond with "warnings". But it's might be not easy to use. Also as mentioned in #1788, using util to wrap warning messages might be helpful. IMHO, controller-runtime should support warning "natively" with `Validator` and `CustomValidator`.

I am interested in this feature/enhancement, so maybe there are possible designs and steps to build this feature:

- extend `Allowed()`, `Denied()`, `ValidationResponse()`, make them support `warning []string`
- bring an interface to extend error, called `ErrorWithWarnings` like:

```go
type ErrorWithWarnings interface {
	error
	Warnings() []string
}
```
- it means it user want to respond with warning, it should return this interface with `ValidateXXX()`
- make type assertion when resolving the error  from `Validator`/`CustomValidator`, fill  `Warnings` in `v1.AdmissionResponse`

Some changes might break the existing API, I think it's OK to controller-runtime. If not, I could still try to only introduce new API without breaking the exist one.

**Comment by STRRL** [comment]
what do you think? If you are not satisfied with this design, I am very happy to make it better. :)

/cc @FillZpp @vincepri @alvaroaleman

**Comment by FillZpp** [comment]
@STRRL Thanks. Does the existing `WithWarnings` function not satisfy the requirement?

https://github.com/kubernetes-sigs/controller-runtime/blob/26c95adfec6390a55f76b2c3ca73ffb532b12dec/pkg/webhook/admission/response.go#L116-L121

We can use it with any response like `Allowed()`, `Denied()` or `ValidationResponse()`. For example

```golang
    // ...
    return admission.Allowed("").WithWarnings("warning message ...")
```

**Comment by STRRL** [comment]
Ohh, I missed that method. That would help a lot. We do not require the first step in my original design. ❤️

But the user still needs to implement `Handler` instead of `Validator` and `CustomValidator`, I think maybe introducing `ErrorWithWarnings` still makes sense.

**Comment by FillZpp** [comment]
Emm... I don't really understand. `Handler` implementation just has to define a func `Handle(context.Context, Request) Response` and its response could also use the `WithWarnings`.

**Comment by STRRL** [comment]
I think it's better also support waning on high-level API `Validator`/`CustomValidator` based on the provided low-level API `Response.WithWarnings`. :)

**Comment by STRRL** [comment]
`ErrorCode`, `Reason` could be carried with `apierrors.APIStatus`:

https://github.com/kubernetes-sigs/controller-runtime/blob/b826f390522092f54fbea9ff18a97ddde4a828b5/pkg/webhook/admission/validator.go#L71-L78

`Warning` is not considered right now, so I think maybe we could use a similar way. :) 

One different thing is a user might want only to respond "warnings" without "error". I realized that my original design might not work well with this situation, because it could not resolve the error is nil.


Maybe make changes on `Validator` interface (or another interface `ValidatorWarning`) are required:

```go
 type ValidatorWarning interface { 
 	runtime.Object 
 	ValidateCreate() (err error, warnings []string) 
 	ValidateUpdate(old runtime.Object) (err error, warnings []string)  
 	ValidateDelete() (err error, warnings []string)  
 }
```

**Comment by STRRL** [comment]
Hi @invidian , I create a PR for resolve this issue, PTAL ❤️

**Comment by STRRL** [comment]
also cc @FillZpp

**Comment by joelanford** [comment]
> I realized that my original design might not work well with this situation, because it could not resolve the error is nil.

We could resolve that right? If the error is an `ErrorWithWarnings` we can use the embedded error as the "real" error, and _that_ one can be `nil` while warnings is not.

If the returned error is not an `ErrorWithWarnings`, then we have the "real" error directly with no warnings.

**Comment by sbueringer** [comment]
@joelanford's suggestion sounds good to me

**Comment by alvaroaleman** [comment]
Hm, after skimming the conversation I am actually in favor of breaking the interface over making it hard to use in the context of warnings by having to return a specific error type. It will be a visible break that is easy to fix.

The only comment I have on the suggestion in https://github.com/kubernetes-sigs/controller-runtime/issues/1896#issuecomment-1122301943 regarding how that would look like is that errors per golang convention should always be the last return parameter.

**Comment by sbueringer** [comment]
Sounds good to me!

**Comment by fabriziopandini** [comment]
> I am actually in favor of breaking the interface over making it hard to use in the context of warnings by having to return a specific error type. It will be a visible break that is easy to fix.

+1

**Comment by STRRL** [comment]
> > I realized that my original design might not work well with this situation, because it could not resolve the error is nil.
> 
> We could resolve that right? If the error is an `ErrorWithWarnings` we can use the embedded error as the "real" error, and _that_ one can be `nil` while warnings is not.
> 
> If the returned error is not an `ErrorWithWarnings`, then we have the "real" error directly with no warnings.

No. I do not think we could resolve that, because `ErrorWithWarnings` is an `interface`(which composes an `error` interface), not a `struct`(which embeds an `error` instance). We could not return a result that carries the "warnings" and  also introduces  "error is nil" at the same time

**Comment by STRRL** [comment]
> Hm, after skimming the conversation I am actually in favor of breaking the interface over making it hard to use in the context of warnings by having to return a specific error type. It will be a visible break that is easy to fix.
> 
> The only comment I have on the suggestion in [#1896 (comment)](https://github.com/kubernetes-sigs/controller-runtime/issues/1896#issuecomment-1122301943) regarding how that would look like is that errors per golang convention should always be the last return parameter.

I would update the PR soon!
